src/constraints.rs

```diff
@@ -0,0 +1,536 @@
+    bits::boolean::Boolean,
+    bits::uint8::UInt8,
+    eq::EqGadget,
+    pairing::PairingVar,
+    ToBitsGadget, ToBytesGadget,
+};
+use ark_relations::r1cs::{Namespace, SynthesisError};
+use ark_std::{borrow::Borrow, marker::PhantomData, vec::Vec};
+
+/// The proof variable for the Groth16 construction
+#[derive(Derivative)]
+#[derivative(Clone(bound = "P::G1Var: Clone, P::G2Var: Clone"))]
+pub struct ProofVar<E: PairingEngine, P: PairingVar<E>> {
+    /// The `A` element in `G1`.
+    pub a: P::G1Var,
+    /// The `B` element in `G2`.
+    pub b: P::G2Var,
+    /// The `C` element in `G1`.
+    pub c: P::G1Var,
+}
+
+/// The verifying key variable for the Groth16 construction
+#[derive(Derivative)]
+#[derivative(
+    Clone(bound = "P::G1Var: Clone, P::GTVar: Clone, P::G1PreparedVar: Clone, \
+    P::G2PreparedVar: Clone, ")
+)]
+pub struct VerifyingKeyVar<E: PairingEngine, P: PairingVar<E>> {
+    /// The `alpha * G`, where `G` is the generator of `E::G1`.
+    pub alpha_g1: P::G1Var,
+    /// The `beta * G`, where `G` is the generator of `E::G2`.
+    pub beta_g2: P::G2Var,
+    /// The `gamma * G`, where `G` is the generator of `E::G2`.
+    pub gamma_g2: P::G2Var,
+    /// The `delta * G`, where `G` is the generator of `E::G2`.
+    pub delta_g2: P::G2Var,
+    /// The `gamma^{-1} * (alpha * a_i + beta * b_i + c_i) * H`, where `H` is the generator of `E::G1`.
+    pub gamma_abc_g1: Vec<P::G1Var>,
+}
+
+impl<E: PairingEngine, P: PairingVar<E>> VerifyingKeyVar<E, P> {
+    /// Prepare the verifying key `vk` for use in proof verification.
+    pub fn prepare(&self) -> Result<PreparedVerifyingKeyVar<E, P>, SynthesisError> {
+        let alpha_g1_pc = P::prepare_g1(&self.alpha_g1)?;
+        let beta_g2_pc = P::prepare_g2(&self.beta_g2)?;
+
+        let alpha_g1_beta_g2 = P::pairing(alpha_g1_pc, beta_g2_pc)?;
+        let gamma_g2_neg_pc = P::prepare_g2(&self.gamma_g2.negate()?)?;
+        let delta_g2_neg_pc = P::prepare_g2(&self.delta_g2.negate()?)?;
+
+        Ok(PreparedVerifyingKeyVar {
+            alpha_g1_beta_g2,
+            gamma_g2_neg_pc,
+            delta_g2_neg_pc,
+            gamma_abc_g1: self.gamma_abc_g1.clone(),
+        })
+    }
+}
+
+/// Preprocessed verification key parameters variable for the Groth16 construction
+#[derive(Derivative)]
+#[derivative(
+    Clone(bound = "P::G1Var: Clone, P::GTVar: Clone, P::G1PreparedVar: Clone, \
+    P::G2PreparedVar: Clone, ")
+)]
+pub struct PreparedVerifyingKeyVar<E: PairingEngine, P: PairingVar<E>> {
+    /// The element `e(alpha * G, beta * H)` in `E::GT`.
+    pub alpha_g1_beta_g2: P::GTVar,
+    /// The element `- gamma * H` in `E::G2`, prepared for use in pairings.
+    pub gamma_g2_neg_pc: P::G2PreparedVar,
+    /// The element `- delta * H` in `E::G2`, prepared for use in pairings.
+    pub delta_g2_neg_pc: P::G2PreparedVar,
+    /// The `gamma^{-1} * (alpha * a_i + beta * b_i + c_i) * H`, where `H` is the generator of `E::G1`.
+    pub gamma_abc_g1: Vec<P::G1Var>,
```

Comment:
Shouldn't this also contain `vk`?

```suggestion
    #[doc(hidden)]
    pub alpha_g1_beta_g2: P::GTVar,
    #[doc(hidden)]
    pub gamma_g2_neg_pc: P::G2PreparedVar,
    #[doc(hidden)]
    pub delta_g2_neg_pc: P::G2PreparedVar,
    #[doc(hidden)]
    pub gamma_abc_g1: Vec<P::G1Var>,
```